src/Std/Data/DHashMap/Lemmas.lean

```diff
@@ -964,4 +965,701 @@ theorem distinct_keys [EquivBEq α] [LawfulHashable α] :
+  sorry
+
+theorem alter_empty_of_none [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)}
+    (h : f none = none) : empty.alter k f = empty := by
+  rw [alter_empty, h]
+
+theorem alter_empty_of_some [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} {v : β k}
+    (h : f none = some v) : empty.alter k f = empty.insert k v := by
+  rw [alter_empty, h]
+
+@[simp]
+theorem isEmpty_alter [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    -- should we use = or ↔? LHS is Bool, RHS is Prop.
+    -- mem_iff_contains suggests ↔.
+    (m.alter k f).isEmpty = ((m.erase k).isEmpty && (f (m.get? k)).isNone) :=
+  Raw₀.isEmpty_alter _ m.2
+
+theorem contains_alter [LawfulBEq α] {k k': α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).contains k' = if k == k' then (f (m.get? k)).isSome else m.contains k' :=
+  Raw₀.contains_alter ⟨_, _⟩ m.2
+
+theorem mem_alter [LawfulBEq α] {k k': α} {f : Option (β k) → Option (β k)} :
+    k' ∈ m.alter k f ↔ if  k == k' then (f (m.get? k)).isSome = true else k' ∈ m := by
+  simp only [mem_iff_contains, contains_alter, beq_iff_eq, Bool.ite_eq_true_distrib]
+
+@[simp]
+theorem contains_alter_self [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).contains k = (f (m.get? k)).isSome := by
+  simp only [contains_alter, beq_self_eq_true, reduceIte]
+
+@[simp]
+theorem mem_alter_self [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    k ∈ m.alter k f ↔ (f (m.get? k)).isSome := by
+  rw [mem_iff_contains, contains_alter_self]
+
+theorem contains_alter_of_beq_eq_false [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (h : (k == k') = false) : (m.alter k f).contains k' = m.contains k' := by
+  simp only [contains_alter, h, Bool.false_eq_true, reduceIte]
+
+theorem mem_alter_of_beq_eq_false [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)}
+    (h : (k == k') = false) : k' ∈ m.alter k f ↔ k' ∈ m := by
+  simp only [mem_iff_contains, contains_alter_of_beq_eq_false, h]
+
+theorem size_alter [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).size =
+    if m.contains k && (f (m.get? k)).isNone then
+      m.size - 1
+    else if m.contains k = false && (f (m.get? k)).isNone then
+      m.size + 1
+    else
+      m.size :=
+  sorry
+
+theorem size_alter' [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).size =
+    m.size
+      - (if m.contains k then 1 else 0)
+      + (if (f (m.get? k)).isSome then 1 else 0) :=
+  sorry
+
+theorem size_alter_le [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).size ≤ m.size + 1 :=
+  sorry
+
+theorem le_size_alter [LawfulBEq α] {k : α} {f : Option (β k) → Option (β k)} :
+    m.size - 1 ≤ (m.alter k f).size :=
+  sorry
+
+theorem get?_alter [LawfulBEq α] {k k' : α} {f : Option (β k) → Option (β k)} :
+    (m.alter k f).get? k' = if h : k == k' then
+      (cast (congrArg (Option ∘ β) (eq_of_beq h)) (f (m.get? k)))
```

Comment:
These verbose casts make me sad, but so far I have nothing to suggest. Happy if anyone has ideas!

Reply:
I feel the same but unfortunately cannot find a better way... I will close this for now, okay?